pkg/component/component.go

```diff
@@ -411,22 +428,105 @@ func (r *RuntimeSpecs) componentForInputType(
 	}
 }
 
+// Returns a list of components that are not grouped, one component for each similar input.
+func (r *RuntimeSpecs) nonGroupedComponentsForInputType(
+	inputType string,
+	output outputI,
+	featureFlags *features.Flags,
+	componentConfig *ComponentConfig,
+) []Component {
+	// Might be less than len(output.inputs[inputType]) if some inputs are disabled.
+	components := make([]Component, 0, len(output.inputs[inputType]))
+
+	inputSpec, componentErr := r.GetInput(inputType)
+
+	for _, input := range output.inputs[inputType] {
+
+		// Units are being mapped to components, so we need a unique ID for each.
+		componentID := fmt.Sprintf("%s-%s-%s", inputType, output.name, input.id)
+
+		var shipperRef *ShipperReference
+		if componentErr == nil {
+			if output.shipperEnabled {
+				var shipperType string
+				shipperType, componentErr = r.getSupportedShipperType(inputSpec, output.outputType)
+
+				if componentErr == nil {
+					// We've found a valid shipper, construct the reference
+					shipperRef = &ShipperReference{
+						ShipperType: shipperType,
+						ComponentID: fmt.Sprintf("%s-%s", shipperType, output.name),
+						// The unit ID of this connection in the shipper is the same as the
+						// input's component id.
+						UnitID: componentID,
+					}
+				}
+			}
+			if shipperRef == nil {
+				// The shipper is disabled or we couldn't find a supported one.
+				if containsStr(inputSpec.Spec.Outputs, output.outputType) {
+					// We found a fallback output, clear componentErr in case it was
+					// set during shipper selection.
+					componentErr = nil
+				} else if componentErr == nil {
+					// This output is unsupported -- set an error if needed, but don't
+					// overwrite an existing error.
+					componentErr = ErrOutputNotSupported
+				}
+			}
+		}
+
+		var units []Unit
+		if input.enabled {
+			// Might be less than 2 if there is a shipper ref and the component has an error.
+			units = make([]Unit, 0, 2)
+			unitID := fmt.Sprintf("%s-unit", componentID)
+			units = append(units, unitForInput(input, unitID))
+			if shipperRef != nil {
+				// Shipper units are skipped if componentErr isn't nil, because in that
+				// case we generally don't have a valid shipper type to base it on.
+				if componentErr == nil {
+					units = append(units, unitForShipperOutput(output, componentID, shipperRef.ShipperType))
+				}
+			} else {
+				units = append(units, unitForOutput(output, componentID))
+			}
+		}
+
+		components = append(components, Component{
+			ID:         componentID,
+			Err:        componentErr,
+			InputSpec:  &inputSpec,
+			InputType:  inputType,
+			OutputType: output.outputType,
+			Units:      units,
+			Features:   featureFlags.AsProto(),
+			Component:  componentConfig.AsProto(),
+			ShipperRef: shipperRef,
+		})
+	}
+
+	return components
```

Comment:
This duplicate lots of code, possible to make this adjustment smaller in the same code above without duplicating all this logic?

Reply:
Refactored code (100% coverage and 0% duplication on new code).